pytorch_lightning/trainer/connectors/callback_connector.py

```diff
@@ -53,6 +53,19 @@ def on_trainer_init(
             progress_bar_refresh_rate, process_position
         )
 
+    def resolve_resume_from_checkpoint(self):
+        if not self._trainer_has_checkpoint_callbacks():
+            return self.trainer.resume_from_checkpoint
+        checkpoint_callbacks = self.trainer.checkpoint_callbacks[0]
+        if os.path.exists(checkpoint_callbacks.best_model_path):
+            resume_from_checkpoint_options = [
+                checkpoint_callbacks.best_model_path,
+                self.trainer.resume_from_checkpoint
+            ]
+            resume_from_checkpoint_options.sort()
+            return resume_from_checkpoint_options[-1]
+        return self.trainer.resume_from_checkpoint
```

Comment:
I don't think this is correct. Why does it even do reloading from the resume_checkpoint while doing testing?

I think to better resolve this issue this should be fixed/refactored correctly:
https://github.com/PyTorchLightning/pytorch-lightning/blob/176735097ab5be9ee21d3e7a3dedc174f3e0dd3f/pytorch_lightning/accelerators/gpu_accelerator.py#L61-L69
since while testing/training setup_training is called, which contains few things that are required both during testing/training and other required during training only(for eg restoring checkpoint).

Reply:
Hey @rohitgr7.

Here was the problem I was trying to resolve.

1. Load a checkpoint from `resume_from_checkpoint `.
2. Fine-tune the model for several epochs which might create a new best_model_path checkpoint
3. When calling `trainer.test()`, it should use the new best_model_path and not `resume_from_checkpoint`.

If I understand properly, you are saying we should skip this restore from `resume_from_checkpoint` in `.test` and load directly from best_model_path ?

Best,
T.C

Reply:
yes, best checkpoint or any other checkpoint, it will be loaded here https://github.com/PyTorchLightning/pytorch-lightning/blob/d1e97a4f114a285349e31e330c7bf8937bc1ee04/pytorch_lightning/trainer/trainer.py#L770-L785, so we can just skip the `.restore` call. Also a few more hooks are called while doing `.test` like `on_pretrain_routine_start` and `on_pretrain_routine_end` since it calls `setup_training` when doing either train/test, which is incorrect too IMO. So I'd suggest to split/refactor `setup_training` a bit.